### ISSUE TITLE
Search does not prioritize exact match when package name contains numbers

### ISSUE THREAD

**Issue by di** [pypa/warehouse#2877]
For example, [a search for `python-miio`](https://pypi.org/search/?q=python-miio) has the `python2-miio` as the first result, likely because the `2` is being tokenized out of the query.

**Comment by gsb-eng** [comment]
@di Any progress on this?  If not, Can I pick this?

**Comment by brainwane** [comment]
@gsb-eng Yes, please go ahead and pick it up! At least within Warehouse, our rule is: If there is not a corresponding pull request for this issue (visibly linked by issue number/PR number), it is up for grabs. So in cases like this, you don't even need to ask. :)

As always, if you have questions along the way as you work on this, please feel free to ask them here, in [`#pypa-dev` on Freenode](https://webchat.freenode.net/?channels=%23pypa-dev), or the [pypa-dev mailing list](https://groups.google.com/forum/#!forum/pypa-dev). Thank you!

**Comment by waseem18** [comment]
@di Could you please help me in providing me more information about this issue?

**Comment by di** [comment]
@waseem18 Unfortunately I don't know much more than you here. We got a report that packages with a number in them could be prioritized higher in search than a package with the exact name as the search query.

However looking at https://pypi.org/search/?q=python-miio now, it seems like `python-miio` has the highest rank now. We will likely need to do some debugging of the elasticsearch query and the results to see why it (at some point) was possible.

**Comment by waseem18** [comment]
Yeah - I was feeling the same. Because now the search queries seems to work as expected.

Will do some debugging and check.

**Comment by di** [comment]
Here's another current example of this happening:

https://pypi.org/search/?q=psycopg2

**Comment by waseem18** [comment]
@di I presume the issue is because of usage of `lowercase` filter in [normalized_name](https://github.com/pypa/warehouse/blob/b66a37d0f5dd5d77730a1bfa427548aa554c40b5/warehouse/packaging/search.py#L29) which is tokenizing `psycopg2` to `psycopg` and `2`

As per Elastic search, 

>  tokenizer breaks text into terms whenever it encounters a character which is not a letter

**Comment by di** [comment]
Sounds like the same thing would happen with package names containing other valid non-alphanumeric characters like `-`, `_`, and `.`. Perhaps we need to tweak that tokenizer a bit.

**Comment by waseem18** [comment]
Yeah - We can have a custom analyzer so that we can avoid tokenizing numbers and non-alphanumeric characters.

**Comment by di** [comment]
@honzakral is able to help with this issue

**Comment by waseem18** [comment]
Yes, it is @selotape 

You can start looking into this issue. Feel free to ping here if you need help with anything.

**Comment by selotape** [comment]
Another example - [https://pypi.org/search/?q=hachoir3](https://pypi.org/search/?q=hachoir3)

**Comment by selotape** [comment]
Hey,

I created a PR with a small fix and a description of my (incomplete) status. @waseem18 , can we take the conversation there?

[https://github.com/pypa/warehouse/pull/4519](https://github.com/pypa/warehouse/pull/4519)

**Comment by yeraydiazdiaz** [comment]
Seems this issue is not happening anymore, [searching for `python-miio`](https://pypi.org/search/?q=python-miio) or [`psycopg2`](https://pypi.org/search/?q=psycopg2) yield the correct packages as first hits on the search.

**Comment by di** [comment]
Seems like this is still an issue:

> I look up 'rfc6266' because I was told about the module and I want to know more about it. Said module appears sixth even though its name matches exactly my query. I look at the sorting criteria and see it’s supposed to be by relevance. I end up wondering how that is not the most relevant result.

**Comment by mgorny** [comment]
The same problem applies to searching for `pep517`: https://pypi.org/search/?q=pep517. The correct result doesn't appear until the second half of the first page.

**Comment by aphedges** [comment]
I encountered this myself recently, and I found multiple instances where it is broken:

- <https://pypi.org/search/?q=Jinja2> returns `Jinja` first and `Jinja2` as second
- <https://pypi.org/search/?q=chardet> returns `chardet2` first and `chardet` second
- <https://pypi.org/search/?q=mypy> returns `mypy1989` first and `mypy` second
- <https://pypi.org/search/?q=pandas> returns `pandas2` first and `pandas` second

However, search is not always wrong. Both <https://pypi.org/search/?q=beautifulsoup4> and <https://pypi.org/search/?q=flake8> work as expected.

**Comment by takluyver** [comment]
More examples:

- [pytools](https://pypi.org/search/?q=pytools) -> `pytools123456789` before `pytools`
- [h5py](https://pypi.org/search/?q=h5py) -> `H.py` before `h5py

**Comment by di** [comment]
Hi folks, I'm locking discussion on this issue -- while I appreciate the intention behind the replies, we don't need more examples of this happening as there are likely thousands, and there's nothing we can do to resolve specific cases.

**Comment by miketheman** [comment]
Searches correctly returning the desired result [post-deploy of boost](https://github.com/pypi/warehouse/pull/19224):

- https://pypi.org/search/?q=python-miio
- https://pypi.org/search/?q=psycopg2
- https://pypi.org/search/?q=hachoir3
- https://pypi.org/search/?q=rfc6266
- https://pypi.org/search/?q=pep517
- https://pypi.org/search/?q=Jinja2
- https://pypi.org/search/?q=chardet
- https://pypi.org/search/?q=mypy
- https://pypi.org/search/?q=pandas
- https://pypi.org/search/?q=beautifulsoup4
- https://pypi.org/search/?q=flake8
- https://pypi.org/search/?q=pytools
- https://pypi.org/search/?q=h5py
- https://pypi.org/search/?q=yolov5

All searches now result the expected results. Closing!
